### PR TITLE
Remove assert 0x260 - Consider disposed as valid and equivalent to disconnected for this codepath

### PR DESCRIPTION
**Pull request by markfields**
## Description

Fixes [AB#6634](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/6634)

Assert 0x260 says `"Disposed should mean disconnected!"`. We see this hit on occasion in our e2e tests with this sequence of events in the logs (in the case I investigated):

Data_eventName | Data_error
-- | --
fluid:telemetry:ContainerRuntime:ContainerRuntimeDisposed |  
fluid:telemetry:Container:ConnectionStateChange_Disconnected |  
fluid:telemetry:DeltaConnection:ClientClosingDeltaConnection |  
fluid:telemetry:Container:noWaitOnDisconnected |  
fluid:telemetry:Container:ContainerClose |  
fluid:telemetry:SummaryManager:EndingSummarizer | 0x260

The second event there would result in this function on `Container` being called, but note it's a no-op if ContainerRuntime is already disposed:

```ts
	private setContextConnectedState(state: boolean, readonly: boolean): void {
		if (this._runtime?.disposed === false) {
			/**
			 * We want to lie to the ContainerRuntime when we are in readonly mode to prevent issues with pending
			 * ops getting through to the DeltaManager.
			 * The ContainerRuntime's "connected" state simply means it is ok to send ops
			 * See https://dev.azure.com/fluidframework/internal/_workitems/edit/1246
			 */
			this.runtime.setConnectionState(state && !readonly, this.clientId);
		}
	}
```

~So the proposed fix is to explicitly transition the ContainerRuntime to Disconnected state at the top of `dispose`.~

So the proposed fix is to allow this state (connected but disposed) and just return `{ shouldSummarize: false}`

### CODE REVIEW

**Comment by markfields** [comment]
Ok as suspected this is a precarious kind of change to make...

Alternatives:
- Remove the assert
- Try updating that setContextConnectionState function  to set state to disconnected even if runtime is disposed.

Thoughts?

**Comment by anthony-murphy** [comment]
the assert looks wrong to me. dispose shouldn't depend on any other state transitions. dispose should just tear everything down

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM

File: packages/runtime/container-runtime/src/summary/summaryManager.ts [hunk 1]
```diff
@@ -186,15 +190,11 @@ export class SummaryManager extends TypedEventEmitter<ISummarizerEvents> impleme
 			return { shouldSummarize: false, stopReason: "notElectedClient" };
 		}
 
-		if (!this.connectedState.connected) {
+		if (!this.connectedState.connected || this.disposed) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summaryManager.ts:193]: this.disposed has already been checked above.

File: packages/runtime/container-runtime/src/summary/summaryManager.ts [hunk 2]
```diff
@@ -238,6 +238,10 @@ export class SummaryManager extends TypedEventEmitter<ISummarizerEvents> impleme
 
 		this.delayBeforeCreatingSummarizer()
 			.then(async (startWithInitialDelay: boolean) => {
+				if (this.disposed) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summaryManager.ts:241]: This is redundant with what's happening inside `getShouldSummarizeState`. I would keep either or. You can update the `stopReason` in that funciton to return `disposed`.

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/summaryManager.ts:241]: It's not.  That's called in many places and needs to protect against disposed.  This needs to ensure it doesn't run if disposed (look closely at usages of line 252's variable)

File: packages/runtime/container-runtime/src/summary/summaryManager.ts (hunk 1 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/summary/summaryManager.ts:193]: Oops thanks

File: packages/runtime/container-runtime/src/summary/summaryManager.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/summary/summaryManager.ts:241]: i think i agree with @agarwal-navin, and we only need this in getShouldSummarizeState. all the code here is synchronous, so exiting earlier doesn't do much

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/summaryManager.ts:241]: If `startWithInitialDelay` is false it won't exit early without this.

**Comment by markfields** [comment]
Added a unit test.  As painful as you would expect, but it was a good exercise in reasoning over async code.  I observed that both disposed checks are required.

**Review by anthony-murphy — APPROVED** [review]

**Pull request merged**